From 49ec54c1e43d8f01adbe99b0df548dae914be2cd Mon Sep 17 00:00:00 2001
From: Yuval Caduri <cyuval@marvell.com>
Date: Sun, 3 Jul 2016 17:37:58 +0300
Subject: [PATCH 0374/1345] axim: fix nr_chan caculcation for AXIM v1

commit  dfa31ecc9bcada7ffc1f93719e88b2c3e2256e68 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

AXI monitor v1 doesn't have the version register so
we can't get the channel count and other info.
When reading from the non-existent register, the v1 monitor
return 0 so we use this behavior to emulate a register read.
the only value that needs to be fixed is the channel count
that equals to 4 for v1

Change-Id: I8c0f74616d7be67592be3bdd5ac2db6817628555
Signed-off-by: Yehuda Yitschak <yehuday@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/31046
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../hwtracing/coresight/coresight-axi-monitor.c    |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/hwtracing/coresight/coresight-axi-monitor.c b/drivers/hwtracing/coresight/coresight-axi-monitor.c
index 8a0a793..df17889 100644
--- a/drivers/hwtracing/coresight/coresight-axi-monitor.c
+++ b/drivers/hwtracing/coresight/coresight-axi-monitor.c
@@ -544,12 +544,21 @@ static void axim_init_default_data(struct axim_drvdata *axim)
 	u32 reg;
 
 	reg = readl_relaxed(axim->base + AXI_MON_VER);
-	axim->nr_chan = BMVAL(reg, 12, 15); /* NCH */
 	axim->nr_prof_reg = BMVAL(reg, 16, 19); /* NPRR */
 	axim->latency_en = BMVAL(reg, 24, 24);
 	axim->trace_en = BMVAL(reg, 25, 25);
 	axim->minor = BMVAL(reg, 4, 7);
 	axim->major = BMVAL(reg, 0, 3) + 1;
+
+	/* AXIM v1 doesn't have a version register and will
+	 * return 0 when reading from AXI_MON_VER. For all fields it
+	 * emulates a VER register fine except for nr_chan
+	 */
+	if (reg)
+		axim->nr_chan = BMVAL(reg, 12, 15); /* NCH */
+	else
+		axim->nr_chan = 4;
+
 }
 
 static int axim_probe(struct amba_device *adev, const struct amba_id *id)
-- 
1.7.9.5

